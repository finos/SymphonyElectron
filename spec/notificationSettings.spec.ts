import { shallow } from 'enzyme';
import * as React from 'react';
import NotificationSettings from '../src/renderer/components/notification-settings';
import { ipcRenderer } from './__mocks__/electron';

describe('Notification Settings', () => {
    const notificationSettingsLabel = 'notification-settings-data';
    const notificationSettingsMock = {
        position: 'upper-right',
        screens: [
            {
                id: '6713899',
            },
            {
                id: '3512909',
            },
        ],
        display: '6713899',
    };
    const onLabelEvent = 'on';
    const sendEvent = 'send';
    const removeListenerLabelEvent = 'removeListener';

    describe('should mount, unmount and render component', () => {
        it('should render the component', () => {
            const wrapper = shallow(React.createElement(NotificationSettings));
            expect(wrapper).toMatchSnapshot();
        });

        it('should call `notification-settings-data` event when component is mounted', () => {
            const spy = jest.spyOn(ipcRenderer, onLabelEvent);
            shallow(React.createElement(NotificationSettings));
            expect(spy).toBeCalledWith(notificationSettingsLabel, expect.any(Function));
        });

        it('should call `updateState` when component is mounted', () => {
            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            shallow(React.createElement(NotificationSettings));

            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            expect(spy).toBeCalledWith(notificationSettingsMock);
        });

        it('should remove listener `notification-settings-data` when component is unmounted', () => {
            const spyMount = jest.spyOn(ipcRenderer, onLabelEvent);
            const spyUnmount = jest.spyOn(ipcRenderer, removeListenerLabelEvent);

            const wrapper = shallow(React.createElement(NotificationSettings));
            expect(spyMount).toBeCalledWith(notificationSettingsLabel, expect.any(Function));

            wrapper.unmount();
            expect(spyUnmount).toBeCalledWith(notificationSettingsLabel, expect.any(Function));
        });
    });

    describe('should select display', () => {
        it('should select display from drop down', () => {
            const notificationSettingsMock = {
                position: 'upper-right',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            const selectDisplaySpy = jest.spyOn(NotificationSettings.prototype, 'selectDisplay');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const positionButton = `select.display-selector`;
            const input = wrapper.find(positionButton);

            input.simulate('change', { target: { value: '6713899' } });

            expect(selectDisplaySpy).toBeCalled();
            expect(spy).toBeCalledWith(notificationSettingsMock);
        });

    });

    describe('should set display position', () => {
        it('should select top right position', () => {
            const notificationSettingsMock = {
                position: 'upper-right',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            const togglePositionButtonSpy = jest.spyOn(NotificationSettings.prototype, 'togglePosition');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const positionButton = `button.upper-right`;
            const input = wrapper.find(positionButton);

            input.simulate('click', { target: { value: 'upper-right' } });

            expect(togglePositionButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith(notificationSettingsMock);
        });

        it('should select bottom right position', () => {
            const notificationSettingsMock = {
                position: 'bottom-right',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            const togglePositionButtonSpy = jest.spyOn(NotificationSettings.prototype, 'togglePosition');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const positionButton = `button.lower-right`;
            const input = wrapper.find(positionButton);

            input.simulate('click', { target: { value: 'lower-right' } });

            expect(togglePositionButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith(notificationSettingsMock);
        });

        it('should select top left position', () => {
            const notificationSettingsMock = {
                position: 'upper-left',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            const togglePositionButtonSpy = jest.spyOn(NotificationSettings.prototype, 'togglePosition');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const positionButton = `button.upper-left`;
            const input = wrapper.find(positionButton);

            input.simulate('click', { target: { value: 'upper-left' } });

            expect(togglePositionButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith(notificationSettingsMock);
        });

        it('should select bottom left position', () => {
            const notificationSettingsMock = {
                position: 'lower-left',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(NotificationSettings.prototype, 'setState');
            const togglePositionButtonSpy = jest.spyOn(NotificationSettings.prototype, 'togglePosition');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const positionButton = `button.lower-left`;
            const input = wrapper.find(positionButton);

            input.simulate('click', { target: { value: 'lower-left' } });

            expect(togglePositionButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith(notificationSettingsMock);
        });
    });

    describe('should submit or cancel new preferences', () => {
        it('should close window on pressing cancel button', () => {
            const notificationSettingsMock = {
                position: 'lower-left',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(ipcRenderer, sendEvent);
            const closeButtonSpy = jest.spyOn(NotificationSettings.prototype, 'close');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const closeButton = `button.footer-button-dismiss`;
            const input = wrapper.find(closeButton);

            input.simulate('click');

            expect(closeButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith('symphony-api', {
                cmd: 'close-window',
                windowType: 'notification-settings',
            });
        });

        it('should submit new preferences on pressing ok button', () => {
            const notificationSettingsMock = {
                position: 'lower-left',
                screens: [],
                display: '6713899',
            };

            const spy = jest.spyOn(ipcRenderer, sendEvent);
            const closeButtonSpy = jest.spyOn(NotificationSettings.prototype, 'close');

            const wrapper = shallow(React.createElement(NotificationSettings));
            ipcRenderer.send('notification-settings-data', notificationSettingsMock);

            const closeButton = `button.footer-button-ok`;
            const input = wrapper.find(closeButton);

            input.simulate('click');

            expect(closeButtonSpy).toBeCalled();
            expect(spy).toBeCalledWith('notification-settings-update', {
                position: 'lower-left',
                display: '6713899',
            });
        });
    });

});
