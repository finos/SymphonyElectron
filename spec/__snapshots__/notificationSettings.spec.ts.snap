// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Settings should mount, unmount and render component should render the component 1`] = `
<div
  className="content"
>
  <header
    className="header"
  >
    <span
      className="header-title"
    >
      Set Notification Position
    </span>
  </header>
  <div
    className="form"
  >
    <label
      className="display-label"
    >
      Show on display
    </label>
    <div
      className="display-container"
      id="screens"
    >
      <select
        className="display-selector"
        id="screen-selector"
        onChange={[Function]}
        title="position"
        value={1}
      />
    </div>
    <label
      className="position-label"
    >
      Position
    </label>
    <div
      className="position-container"
    >
      <div
        className="button-set-left"
      >
        <div
          className="position-button-container"
        >
          <button
            className="position-button upper-left"
            id="upper-left"
            name="position"
            onClick={[Function]}
            type="button"
            value="upper-left"
          >
            Top Left
          </button>
        </div>
        <div
          className="position-button-container"
        >
          <button
            className="position-button lower-left"
            id="lower-left"
            name="position"
            onClick={[Function]}
            type="button"
            value="lower-left"
          >
            Bottom Left
          </button>
        </div>
      </div>
      <div
        className="button-set-right"
      >
        <div
          className="position-button-container"
        >
          <button
            className="position-button position-button-selected upper-right"
            id="upper-right"
            name="position"
            onClick={[Function]}
            type="button"
            value="upper-right"
          >
            Top Right
          </button>
        </div>
        <div
          className="position-button-container"
        >
          <button
            className="position-button lower-right"
            id="lower-right"
            name="position"
            onClick={[Function]}
            type="button"
            value="lower-right"
          >
            Bottom Right
          </button>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="footer"
  >
    <div
      className="footer-button-container"
    >
      <button
        className="footer-button footer-button-dismiss"
        id="cancel"
        onClick={[Function]}
      >
        CANCEL
      </button>
      <button
        className="footer-button footer-button-ok"
        id="ok-button"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </footer>
</div>
`;
